src/detectors/builtin/suspiciousLoop.ts

```diff
@@ -0,0 +1,132 @@
+import { CompilationUnit } from "../../internals/ir";
+import {
+  foldStatements,
+  foldExpressions,
+  evalExpr,
+} from "../../internals/tact";
+import { MistiTactWarning, Severity } from "../../internals/warnings";
+import { AstDetector } from "../detector";
+import {
+  AstStatement,
+  AstExpression,
+} from "@tact-lang/compiler/dist/grammar/ast";
+
+/**
+ * SuspiciousLoop Detector
+ *
+ * An optional detector that identifies potentially problematic loops, such as those
+ * with unbounded conditions or excessive iteration counts.
+ *
+ * ## Why is it bad?
+ * Loops with always-true conditions or massive iteration limits can lead to high
+ * gas consumption and even denial of service (DoS) issues. By flagging these loops,
+ * this detector aids auditors in catching potential performance or security risks.
+ *
+ * ## Example
+ * ```tact
+ * repeat (1_000_001) { // Highlighted by detector as high iteration count
+ *     // ...
+ * }
+ *
+ * while (true) { // Highlighted as unbounded condition
+ *     // ...
+ * }
+ * ```
+ */
+export class SuspiciousLoop extends AstDetector {
+  severity = Severity.HIGH;
+
+  async check(cu: CompilationUnit): Promise<MistiTactWarning[]> {
+    const processedLoopIds = new Set<number>();
+    return Array.from(cu.ast.getProgramEntries()).reduce((acc, node) => {
+      return acc.concat(
+        ...foldStatements(
+          node,
+          (acc, stmt) => {
+            return acc.concat(
+              this.analyzeLoopStatement(stmt, processedLoopIds),
+            );
+          },
+          acc,
+          { flatStmts: true },
+        ),
+      );
+    }, [] as MistiTactWarning[]);
+  }
+
+  /**
+   * Analyzes a loop statement to determine if it contains a suspicious condition.
+   * @param stmt - The statement to analyze.
+   * @param processedLoopIds - A set of loop IDs already processed.
+   * @returns An array of MistiTactWarning objects if a suspicious loop is detected.
+   */
+  private analyzeLoopStatement(
+    stmt: AstStatement,
+    processedLoopIds: Set<number>,
+  ): MistiTactWarning[] {
+    if (processedLoopIds.has(stmt.id)) {
+      return [];
+    }
+    if (this.isLoop(stmt)) {
+      processedLoopIds.add(stmt.id);
+      return foldExpressions(
+        stmt,
```

Comment:
We should handle only `stmt.condition` for while/until loops or `stmt.iterations` for repeat loops, and skip foreach loops here.

Reply:
For while/until loops we should take into account only expressions evaluable to `false`. For repeat loops we are interested in large numbers appearing after consteval.

Reply:
It was not fixed.